src/function_lua.c

```diff
@@ -224,54 +224,141 @@ static void luaEngineFreeFunction(void *engine_ctx, void *compiled_function) {
+
+            *lua_f_ctx = zmalloc(sizeof(*lua_f_ctx));
+            (*lua_f_ctx)->lua_function_ref = lua_function_ref;
+            continue; /* value was already popped, so no need to pop it out. */
+        } else {
+            /* unknown argument was given, raise an error */
+            err = "unknown argument given to redis.register_function";
+            goto error;
+        }
+        lua_pop(lua, 1); /* pop the value to continue the iteration */
     }
 
-    if (!lua_isstring(lua, 1)) {
-        luaPushError(lua, "first argument to redis.register_function must be a string");
-        return luaRaiseError(lua);
+    if (!*name) {
+        err = "redis.register_function must get a function name argument";
+        goto error;
     }
 
-    if (!lua_isfunction(lua, 2)) {
-        luaPushError(lua, "second argument to redis.register_function must be a function");
-        return luaRaiseError(lua);
+    if (!*lua_f_ctx) {
+        err = "redis.register_function must get a callback argument";
+        goto error;
     }
 
-    if (argc == 3 && !lua_isstring(lua, 3)) {
-        luaPushError(lua, "third argument to redis.register_function must be a string");
-        return luaRaiseError(lua);
+    return C_OK;
+
+error:
+    if (*name) sdsfree(*name);
+    if (*desc) sdsfree(*desc);
+    if (*lua_f_ctx) {
+        lua_unref(lua, (*lua_f_ctx)->lua_function_ref);
+        zfree(*lua_f_ctx);
     }
+    luaPushError(lua, err);
+    return C_ERR;
+}
 
-    size_t function_name_len;
-    const char *function_name = lua_tolstring(lua, 1, &function_name_len);
-    sds function_name_sds = sdsnewlen(function_name, function_name_len);
+static int luaRegisterFunctionReadPositionalArgs(lua_State *lua, luaFunctionCtx **lua_f_ctx, sds *name, sds *desc) {
+    int argc = lua_gettop(lua);
+    char *err = NULL;
+    if (!(*name = luaGetStringSds(lua, 1))) {
+        err = "first argument to redis.register_function must be a string";
+        goto error;
+    }
 
-    sds desc_sds = NULL;
-    if (argc == 3){
-        size_t desc_len;
-        const char *desc = lua_tolstring(lua, 3, &desc_len);
-        desc_sds = sdsnewlen(desc, desc_len);
+    if (!lua_isfunction(lua, 2)) {
+        err = "second argument to redis.register_function must be a function";
+        goto error;
+    }
+
+    if (argc == 3) {
+        if (!(*desc = luaGetStringSds(lua, 3))) {
+            err = "third argument to redis.register_function must be a string";
+            goto error;
+        }
         lua_pop(lua, 1); /* pop out the description */
     }
 
     int lua_function_ref = luaL_ref(lua, LUA_REGISTRYINDEX);
 
-    luaFunctionCtx *lua_f_ctx = zmalloc(sizeof(*lua_f_ctx));
-    *lua_f_ctx = (luaFunctionCtx ) { .lua_function_ref = lua_function_ref, };
+    *lua_f_ctx = zmalloc(sizeof(*lua_f_ctx));
+    (*lua_f_ctx)->lua_function_ref = lua_function_ref;
+
+    return C_OK;
+
+error:
+    if (*name) sdsfree(*name);
+    if (*desc) sdsfree(*desc);
```

Comment:
This seems a bit weird, although we free it, but passed an invalid address to the caller.

Reply:
User should only use those if the function returns C_OK. But I agree its safer to set them to NULL, will change that.

Reply:
Fixed